src/driver/shared_fd.rs

```diff
@@ -74,61 +93,68 @@ impl Inner {
         let state = RefCell::get_mut(&mut self.state);
 
         // Submit a close operation
-        *state = match Op::close(self.fd) {
-            Ok(op) => State::Closing(op),
-            Err(_) => {
-                // Submitting the operation failed, we fall back on a
-                // synchronous `close`. This is safe as, at this point, we
-                // guarantee all in-flight operations have completed. The most
-                // common cause for an error is attempting to close the FD while
-                // off runtime.
-                //
-                // This is done by initializing a `File` with the FD and
-                // dropping it.
-                //
-                // TODO: Should we warn?
+        // If either:
+        //  - runtime has already closed, or
+        //  - submitting the Close operation fails
+        // we fall back on a synchronous `close`. This is safe as, at this point,
+        // we guarantee all in-flight operations have completed. The most
+        // common cause for an error is attempting to close the FD while
+        // off runtime.
+        //
+        // This is done by initializing a `File` with the FD and
+        // dropping it.
+        //
+        // TODO: Should we warn?
+        *state = match CONTEXT.try_with(|cx| cx.is_set()) {
+            Ok(true) => match Op::close(self.fd) {
```

Comment:
As Close is special cased (preserving the current  behavior to immediately submitting to the ring) , we can fallback to synchronous Close as required.